src/compiler/checker.ts

```diff
@@ -1205,7 +1205,7 @@ export const enum TypeFacts {
     FalseStrictFacts = BaseBooleanStrictFacts | Falsy,
     FalseFacts = BaseBooleanFacts,
     TrueStrictFacts = BaseBooleanStrictFacts | Truthy,
-    TrueFacts = BaseBooleanFacts | Truthy,
+    TrueFacts = (BaseBooleanFacts & ~Falsy) | Truthy,

```

Comment:
if this is too broad of a change then perhaps the fix (if it's even desired today) could be moved closer to this `narrowType` callback:
https://github.dev/microsoft/TypeScript/blob/e936eb13d2900f21d79553c32a704307c7ad03dd/src/compiler/checker.ts#L27121